refactor: gragp & workflow

meguminnnnnnnnn opened the pull request:
#### What type of PR is this?
<!--
Add one of the following kinds:

build: Changes that affect the build system or external dependencies (example scopes: gulp, broccoli, npm)
ci: Changes to our CI configuration files and scripts (example scopes: Travis, Circle, BrowserStack, SauceLabs)
docs: Documentation only changes
feat: A new feature
optimize: A new optimization
fix: A bug fix
perf: A code change that improves performance
refactor: A code change that neither fixes a bug nor adds a feature
style: Changes that do not affect the meaning of the code (white space, formatting, missing semi-colons, etc)
test: Adding missing tests or correcting existing tests
chore: Changes to the build process or auxiliary tools and libraries such as documentation generation
-->

#### Check the PR title.
<!--
The description of the title will be attached in Release Notes, 
so please describe it from user-oriented, what this PR does / why we need it.
Please check your PR title with the below requirements:
-->
- [ ] This PR title match the format: \<type\>(optional scope): \<description\>
- [ ] The description of this PR title is user-oriented and clear enough for others to understand.
- [ ] Attach the PR updating the user documentation if the current PR requires user awareness at the usage level. [User docs repo](https://github.com/cloudwego/cloudwego.github.io)


#### (Optional) Translate the PR title into Chinese.


#### (Optional) More detailed description for this PR(en: English/zh: Chinese).
<!--
Provide more detailed info for review(e.g., it's recommended to provide perf data if this is a perf type PR).
-->
en:
zh(optional): 


#### (Optional) Which issue(s) this PR fixes:
<!--
Automatically closes linked issue when PR is merged.
Eg: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
-->

#### (optional) The PR that updates user documentation:
<!--
If the current PR requires user awareness at the usage level, please submit a PR to update user docs. [User docs repo](https://github.com/cloudwego/cloudwego.github.io)
-->

shentongmartin commented on compose/graph_run.go:
```diff
@@ -165,253 +109,206 @@ func (r *runner) transform(ctx context.Context, input streamReader, opts ...Opti
 		output:  input,
 	})
 
 	for step := 0; ; step++ {
```
eager 模式下的 step 该怎么理解，就不再是 super step 了，而是节点数。WithMaxStep 的具体含义会根据 eager 与否而有区别

meguminnnnnnnnn replied:
是这样，这个看需不需要改吧，我是感觉还好

shentongmartin commented on compose/graph_run.go:
```diff
@@ -165,253 +109,206 @@ func (r *runner) transform(ctx context.Context, input streamReader, opts ...Opti
-				return nil, fmt.Errorf("write_to_channel (node): %s not present in the graph", wCh)
+			if v, ok := nodeMap[END]; ok {
+				// reach end
+				return v, nil
```
需要有个资源释放的过程，比如 TaskManager 里面的 chan

meguminnnnnnnnn replied:
chan不需要手动释放吧

shentongmartin commented on compose/graph_manager.go:
```diff
@@ -0,0 +1,184 @@
+	err     error
+}
+
+type taskManager struct {
```
结构如果改一下是否会简单：去掉 list，去掉 mutex，保留 channel 和 num。submit 就是增加计数，异步往 channel 塞 task。WaitOne 就是从 channel 里取一个 task，计数减一。WaitAll 就是从 channel 里取出 num 个 task。submit 和 wait 应该都没有并发，好像不需要上锁，可以直接操作 num

meguminnnnnnnnn replied:
上锁是为了操作list，不是操作num； 加list是为了task完成后不会被channel阻塞，任务执行完goroutine就可以退出

shentongmartin replied:
channel 给个 buffer 是不是也基本平替了，buffer 就是那个 list 的作用？

meguminnnnnnnnn replied:
需要的buffer长度基本正比于graph并发度，graph里没有相关指标指示怎么配置buffer长度

shentongmartin commented on compose/field_mapping.go:
```diff
@@ -0,0 +1,429 @@
+		return nil, fmt.Errorf("invalid field mappings: from all fields to all, use common edge instead")
+	} else if !isToAll(mappings) && validateStruct(outType) {
+		// 如果outType非struct要报错，因为不是toAll时，outType必须是struct，否则没办法构造
+		return nil, fmt.Errorf("static check fail: upstream input type should be struct, actual: %v", outType)
```
报错信息反了？这里校验的是下游的输入类型？

meguminnnnnnnnn replied:
没反，命名问题，改了

shentongmartin commented on compose/field_mapping.go:
```diff
@@ -0,0 +1,429 @@
+func validateFieldMapping(inType reflect.Type, outType reflect.Type, mappings []*FieldMapping) (map[FieldMapping]func(any) (any, error), error) {
+	var fieldCheckers = make(map[FieldMapping]func(any) (any, error))
+
+	// 检查 FromAll 到 ToAll 的情况
```
这里是不是重复了，下面的 checkAndExtractFieldType 涵盖了大部分情况？

meguminnnnnnnnn replied:
谁和谁重复了

shentongmartin replied:
checkAndExtractFieldType 里面，有对于是不是 struct 的校验

meguminnnnnnnnn replied:
```go
if len(field) == 0 {
	return typ, nil
}
for typ.Kind() == reflect.Ptr {
	typ = typ.Elem()
}

if typ.Kind() != reflect.Struct {
	return nil, fmt.Errorf("type[%v] is not a struct", typ)
}
```
checkAndExtractFieldType的功能是获得inputType指定字段的type，加了一个特殊处理，如果指定字段为空就返回自己的type，只会报程序性错误，比如要从string里取一个字段

validate校验的是mapping配置， 下游要映射字段但是提供了一个非struct，和上游要映射字段但是提供了一个非struct。后者可以runtime校验，但是有点不太合理，用户声明了一个interface，然后在field mapping里指定了一个字段？如果要指定字段就应该用struct，如果要用interface，那就应该依赖interface的方法，而不是某个interface实例的字段，如果指定一个interface的某个字段，感觉不是很好的做法，业务代码这样写也不合理。

shentongmartin replied:
mapping 配置中，如果下游要映射字段，但是提供了一个非 struct，用 checkAndExtractFieldType 也是能校验出来的？

meguminnnnnnnnn replied:
好像是可以...

meguminnnnnnnnn replied:
但是为啥我感觉加上更清晰点

meguminnnnnnnnn replied:
要不还是留着吧，更明显一点。。

shentongmartin commented on compose/field_mapping.go:
```diff
@@ -0,0 +1,429 @@
+			fieldCheckers[*mapping] = func(a any) (any, error) {
+				trueInType := reflect.TypeOf(a)
+				if !trueInType.AssignableTo(outType) {
+					return nil, fmt.Errorf("runtime check failed for mapping %s, field[%v]-[%v] must not be assignable", mapping, inField, outField)
```
error 信息有点问题？

meguminnnnnnnnn replied:
done

meguminnnnnnnnn commented:
> ## ⚠️ Breaking API Changes Detected
> This pull request contains breaking API changes that may affect compatibility:
> 
> ```
> uncompatible api changes found: 
> github.com/cloudwego/eino/components/retriever
>   Incompatible changes:
>   - GetImplSpecificOptions: removed
>   - WrapImplSpecificOptFn: removed
> 
> github.com/cloudwego/eino/components/embedding
>   Incompatible changes:
>   - GetImplSpecificOptions: removed
>   - WrapImplSpecificOptFn: removed
> 
> github.com/cloudwego/eino/components/indexer
>   Incompatible changes:
>   - GetImplSpecificOptions: removed
>   - WrapImplSpecificOptFn: removed
> 
> github.com/cloudwego/eino/compose
>   Compatible changes:
>   - ComponentOfWorkflow: added
>   - FieldMapping: added
>   - NewMapping: added
>   - NewWorkflow: added
>   - Workflow: added
>   - WorkflowNode: added
> ```
> 
> ### Required Actions:
> 1. Please review these changes carefully
> 2. Update the user documentation to reflect these API changes
> 3. Add a note in the changelog about these breaking changes

meguminnnnnnnnn commented on compose/workflow_test.go:
```diff
@@ -0,0 +1,427 @@
+		assert.Equal(t, []*schema.Message{{Role: schema.User, Content: "query_query"}}, out)
+	})
+
+	// 没有扇入的类型检测
```
之前的graph就没有这个，扇入只检查了类型相同

shentongmartin replied:
graph 不好检查，workflow 的扇入比较明确，好检查？或者说我们一定要保持 workflow 的校验逻辑和 graph 的校验逻辑相同吗

meguminnnnnnnnn replied:
保持一致比较好吧，搞那么多差异我们自己都搞不明白

meguminnnnnnnnn replied:
这个和"a node has no input"那个看一下graph需不需要加吧，需要加就加上

meguminnnnnnnnn commented on compose/workflow_test.go:
```diff
@@ -0,0 +1,427 @@
+		assert.ErrorContains(t, err, "edge start node '2' needs to be added to graph first")
+	})
+
+	//t.Run("multiple mappings have an empty mapping", func(t *testing.T) {
```
相同dest不是改成merge了

shentongmartin replied:
这里是相同的 FromNodeKey，不是不同的 FromNodeKey，不同的确实是 merge

meguminnnnnnnnn replied:
相同的from node key不允许to同一个字段吗，为啥有这个限制

shentongmartin replied:
好像也行，我的直觉认为这是配错了。。

meguminnnnnnnnn commented on compose/workflow_test.go:
```diff
@@ -0,0 +1,427 @@
+	ctx := context.Background()
+	ctrl := gomock.NewController(t)
+
+	//t.Run("a node has no input", func(t *testing.T) {
```
有个node 没有前序， 这个也是原本graph没有的检查

shentongmartin replied:
graph 不好检查，但是 workflow 检查一下成本不高？

meguminnnnnnnnn replied:
为啥会锁住，节点没有前序那这个节点永远不会被执行，如果没有节点可执行了会报错no task to execute 然后退出啊

shentongmartin replied:
ok

meguminnnnnnnnn commented on compose/workflow.go:
```diff
@@ -0,0 +1,185 @@
+}
+
+func (wf *Workflow[I, O]) AddChatModelNode(key string, chatModel model.ChatModel, opts ...GraphAddNodeOpt) *WorkflowNode {
+	_ = wf.g.AddChatModelNode(key, chatModel, opts...)
```
记下来有啥用，打日志吗

meguminnnnnnnnn commented on compose/workflow.go:
```diff
@@ -0,0 +1,185 @@
+	for _, input := range inputs {
+		input.fromNodeKey = fromNodeKey
+	}
+	_ = wf.g.addEdgeWithMappings(fromNodeKey, END, inputs...)
```
会

meguminnnnnnnnn commented on compose/state.go:
```diff
@@ -116,7 +121,25 @@ func streamConvertPostHandler[O, S any](handler StreamStatePostHandler[O, S]) *c
+	return cState, nil
+}
+
+func getState[S any](ctx context.Context) (S, error) {
```
这个是给eager用的，还是待定的。
小写getState无论如何都会成功，graph调用pre、post handler时用这个方法，eager模式里也可以安全调用这个。
大写在eager模式会报错，因为eager模式node代码和pre、post handler有并发，所以不允许node代码里get state了

meguminnnnnnnnn commented on compose/dag.go:
```diff
@@ -65,14 +65,11 @@ func (ch *dagChannel) get(ctx context.Context) (any, error) {
 	return ch.value != nil
 }
-
-func (ch *dagChannel) clear(ctx context.Context) {
```
get完一定clear，就不需要单独clear了

meguminnnnnnnnn commented on compose/state.go:
```diff
@@ -30,6 +30,11 @@ type GenLocalState[S any] func(ctx context.Context) (state S)
 
+type internalState struct {
+	state any
+	eager bool
```
没有，eager现在只是graph运行支持了，api还没定

meguminnnnnnnnn commented on compose/field_mapping.go:
```diff
@@ -0,0 +1,429 @@
+}
+
+// NewMapping creates a new FieldMapping with the specified fromNodeKey.
+func NewMapping() *FieldMapping {
```
这样体现不出from和to是可选的， 虽然现在这个链式的感觉也不太好。。

shentongmartin replied:
提供三个：
1. func From(fieldName) *Mapping
2. func (m *Mapping) To(fieldName)
3. func To(fieldName) *Mapping

meguminnnnnnnnn commented on compose/state.go:
```diff
@@ -116,7 +121,25 @@ func streamConvertPostHandler[O, S any](handler StreamStatePostHandler[O, S]) *c
+	iState := state.(*internalState)
+	if iState.eager {
+		var s S
+		return s, fmt.Errorf("GetState in node is forbidden in eager mode")
```
对，暂时没想到更好的办法

meguminnnnnnnnn commented on compose/field_mapping.go:
```diff
@@ -0,0 +1,429 @@
+	return output, false, nil
+}
+
+func newFieldMappingHandler() *fieldMappingHandler {
```
删掉了

meguminnnnnnnnn commented on compose/field_mapping.go:
```diff
@@ -0,0 +1,429 @@
+type fieldMappingHandler struct {
+	invoke      func(output any) (map[FieldMapping]any, error)
+	stream      func(output streamReader) *schema.StreamReader[map[FieldMapping]any]
+	preCheckers func(value any) (any, error)
```
删掉了

meguminnnnnnnnn commented on compose/field_mapping.go:
```diff
@@ -0,0 +1,429 @@
+
+type fieldMappingHandler struct {
+	invoke      func(output any) (map[FieldMapping]any, error)
+	stream      func(output streamReader) *schema.StreamReader[map[FieldMapping]any]
```
done

meguminnnnnnnnn commented on compose/field_mapping.go:
```diff
@@ -0,0 +1,429 @@
+				if checker, ok := f.checkers[k]; ok {
+					nValue, checkErr := checker(v)
+					if checkErr != nil {
+						return nil, checkErr
```
done

meguminnnnnnnnn commented on compose/graph.go:
```diff
@@ -150,20 +151,30 @@ type graph struct {
 	runtimeCheckEdges    map[string]map[string]bool
 	runtimeCheckBranches map[string][]bool
 
+	fieldMappings *fieldMappingsOnEdge
```
加过了，更新下

meguminnnnnnnnn commented on compose/graph.go:
```diff
@@ -543,15 +584,18 @@ func (g *graph) AddBranch(startNode string, branch *GraphBranch) (err error) {
 	if startNodeOutputType == nil && endNodeInputType == nil {
 		// type of passthrough have not been inferred yet. defer checking to compile.
-		g.toValidateMap[startNode] = append(g.toValidateMap[startNode], endNode)
+		g.toValidateMap[startNode] = append(g.toValidateMap[startNode], struct {
```
小graph是支持passthrough的filedmapping的，

meguminnnnnnnnn replied:
顺手加上了， Workflow可以不暴露

meguminnnnnnnnn commented on compose/graph_manager.go:
```diff
@@ -0,0 +1,184 @@
+	for target, v := range values {
+		toChannel, ok := (*c)[target]
+		if !ok {
+			return fmt.Errorf("target channel isn't existed: %s", target)
```
要是branch输出了一个其他node key会走到这里

meguminnnnnnnnn commented on compose/graph_manager.go:
```diff
@@ -0,0 +1,184 @@
+	return nil
+}
+
+func (c *channelManager) getReady(ctx context.Context) (map[string]any, error) {
```
done

meguminnnnnnnnn commented on compose/graph_manager.go:
```diff
@@ -0,0 +1,184 @@
+					currentTask.err = safe.NewPanicErr(panicInfo, debug.Stack())
+				}
+				t.mu.Lock()
+				t.l.PushBack(currentTask)
```
taskManager里的链表并发不安全，需要上锁保护。  pushback是加入到链表末尾。这里是先上锁，然后task入队，然后尝试队头元素加入channel，开锁

meguminnnnnnnnn commented on compose/graph.go:
```diff
@@ -271,57 +319,50 @@ func (g *graph) addNode(key string, node *graphNode, options *graphAddNodeOpts)
+	return g.addEdgeWithMappings(startNode, endNode)
+}
+
+func (g *graph) addEdgeWithMappings(startNode, endNode string, mappings ...*FieldMapping) (err error) {
```
对

shentongmartin commented on compose/workflow_test.go:
```diff
@@ -0,0 +1,371 @@
+		}),
+		WithInputKey("key"),
+		WithOutputKey("key1")).
+		AddInput("1") // map["key"][]any ->> map[""]map["key"][]any -> map["key"][]any -> []any -> map["key1"][]any
```
注释改下或者删掉吧，已经过时了，key 不是 field，是 mapping 了

meguminnnnnnnnn replied:
done

shentongmartin replied:
下面还有其他的同类注释

meguminnnnnnnnn replied:
应该也改了吧

shentongmartin commented on compose/graph_manager.go:
```diff
@@ -0,0 +1,184 @@
+func (t *taskManager) updateChan() {
+	for t.l.Len() > 0 {
+		select {
+		case t.done <- t.l.Front().Value.(*task):
```
这里往 t.done 里面塞，是阻塞的等待 waitOne 从里面取？如果是这样的话，好像 goroutine 的释放也是需要依赖 waitOne 从 channel 里取？这样跟去掉 list，直接往 channel 里发的区别是什么

meguminnnnnnnnn replied:
select有default，是非阻塞的

> 在Go语言中，`select`语句用于处理通道（channel）的并发操作，而`select default`则是`select`语句中的一种特殊情况。`select default`语法表示在没有任何通道准备就绪时，立即执行的默认情况。
> 当使用`select`语句监听多个通道时，如果没有任何通道准备就绪，程序会阻塞在`select`语句处等待通道的准备。但如果在`select`语句中包含了`default`分支，即`select`` default`，那么当没有任何通道准备就绪时，程序会立即执行`default`分支的代码块，而不会阻塞等待通道。

shentongmartin replied:
select 不阻塞，但是 for t.l.Len() > 0 这个是阻塞的吧，updateChan 这个方法整体是依赖主线程的 wait 的，间接导致每个 task 的 goroutine 的释放其实是依赖主线程的 wait 的

meguminnnnnnnnn replied:
default就直接return了.....

shentongmartin replied:
ok

shentongmartin commented on compose/graph_run.go:
```diff
@@ -75,25 +77,20 @@ type runner struct {
+	inputConverter              handlerPair
+	inputFieldMappingConverter  handlerPair
+	outputConverter             handlerPair
+	outputFieldMappingConverter handlerPair
```
outputFieldMappingConverter 看起来没用到？

meguminnnnnnnnn replied:
删掉了

meguminnnnnnnnn commented on compose/graph_run.go:
```diff
@@ -447,14 +355,12 @@ func (r *runner) toComposableRunnable() *composableRunnable {
+		outputType:                 r.outputType,
+		inputStreamFilter:          r.inputStreamFilter,
+		inputConverter:             r.inputConverter,
+		inputFieldMappingConverter: r.inputFieldMappingConverter,
```
inputxxx都只是传递，不是runner用的

meguminnnnnnnnn commented on compose/workflow.go:
```diff
@@ -34,6 +33,7 @@ import (
 type WorkflowNode struct {
 	g   *graph
 	key string
+	err error
```
删掉了

meguminnnnnnnnn commented on compose/workflow_test.go:
```diff
@@ -334,16 +334,13 @@ func TestFanInToSameDest(t *testing.T) {
+		}), WithOutputKey("B")).AddInput(START, From("B"))
+		wf.AddEnd("1", To("F")).AddEnd("2", To("F"))
+		_, err := wf.Compile(context.Background())
+		assert.Error(t, err)
```
done

meguminnnnnnnnn commented on compose/chain.go:
```diff
@@ -121,6 +121,20 @@ func (c *Chain[I, O]) addEndIfNeeded() error {
+		invoke:    buildFieldMappingConverter[I](),
+		transform: buildStreamFieldMappingConverter[I](),
+	}
+}
```
chain做子图的时候需要

meguminnnnnnnnn commented on compose/types_composable.go:
```diff
@@ -26,5 +26,7 @@ type AnyGraph interface {
 	inputType() reflect.Type
 	outputType() reflect.Type
+	inputConverter() handlerPair
+	inputFieldMappingConverter() handlerPair
```
确实不太对，但这个有需要。。
